core/solver/multigrid.cpp

```diff
@@ -0,0 +1,595 @@
+            d_list.emplace_back(vec::create(exec, vector_size));
+            // 1 x nrhs norm_vec
+            old_norm_list.emplace_back(norm_vec::create(exec, scaler_size));
+            new_norm_list.emplace_back(norm_vec::create(exec, scaler_size));
+        }
+    }
+
+    void run_cycle(multigrid_cycle cycle, size_type level,
+                   const std::shared_ptr<const LinOp> &matrix, const LinOp *b,
+                   LinOp *x)
+    {
+        if (level == multigrid->get_mg_level_list().size()) {
+            multigrid->get_coarsest_solver()->apply(b, x);
+            return;
+        }
+        auto mg_level = multigrid->get_mg_level_list().at(level);
+        run<gko::multigrid::EnableMultigridLevel, float, double,
+            std::complex<float>, std::complex<double>>(
+            mg_level, [&, this](auto mg_level) {
+                using value_type = typename std::decay_t<
+                    detail::pointee<decltype(mg_level)>>::value_type;
+                this->run_cycle<value_type>(cycle, level, matrix, b, x);
+            });
+    }
+
+    template <typename VT>
+    void run_cycle(multigrid_cycle cycle, size_type level,
+                   const std::shared_ptr<const LinOp> &matrix, const LinOp *b,
+                   LinOp *x)
+    {
+        auto total_level = multigrid->get_mg_level_list().size();
+#define as_vec(x) std::static_pointer_cast<matrix::Dense<VT>>(x)
+#define as_real_vec(x) \
+    std::static_pointer_cast<matrix::Dense<remove_complex<VT>>>(x)
+
+        auto r = as_vec(r_list.at(level));
+        auto g = as_vec(g_list.at(level));
+        auto e = as_vec(e_list.at(level));
+        // get mg_level
+        auto mg_level = multigrid->get_mg_level_list().at(level);
+        // get the pre_smoother
+        auto pre_smoother = multigrid->get_pre_smoother_list().at(level);
+        // get the mid_smoother
+        auto mid_smoother = multigrid->get_mid_smoother_list().at(level);
+        // get the post_smoother
+        auto post_smoother = multigrid->get_post_smoother_list().at(level);
+        auto one = one_list.at(level).get();
+        auto neg_one = neg_one_list.at(level).get();
+        // Smoother * x = r
+        if (pre_smoother) {
+            pre_smoother->apply(b, x);
+            // compute residual
+            r->copy_from(b);  // n * b
+            matrix->apply(neg_one, x, one, r.get());
+        } else if (level != 0) {
+            // move the residual computation at level 0 to out-of-cycle if there
+            // is no pre-smoother at level 0
+            r->copy_from(b);
+            matrix->apply(neg_one, x, one, r.get());
+        }
+        // first cycle
+        mg_level->get_restrict_op()->apply(r.get(), g.get());
+        // next level
+        this->run_cycle(cycle, level + 1, mg_level->get_coarse_op(), g.get(),
+                        e.get());
+        // additional work for non-v_cycle
+        if (cycle == multigrid_cycle::f || cycle == multigrid_cycle::w) {
```

Comment:
from wiki, W cycle seems to two cycle on the finest level?
https://en.wikipedia.org/wiki/Multigrid_method

Reply:
But if you look at any multigrid book (say Trottenberg or even Briggs), the W cycle corresponds to "gamma" = 2. That means that the recursive multigrid call on any coarser level is called twice. So on the finest level you call it only once.

Reply:
okay, I also search some book and they also denote the W cycle as you said.
I will change its behavior.
Do you think the following make sense?
I will delete the mid-smoother customization because W-cycle will not use the first mid-smoother anymore.
only F-cycle will use it.
only left the mid-smoother choose pre-smoother or post-smoother option.

Reply:
Yeah, I think that would be fine. I mean you could still choose to keep the option of a separate mid-smoother, but I don't think it's important.

Reply:
after reading one book, the mid smoother does not needed because they do not have prolong and restriction between two multigrid call

Reply:
So what would you do in place of line 293 below, then?

Reply:
I update the behavior and also gives the reference in the cycle.
please take a look